scipy/cluster/hierarchy.py

```diff
@@ -3855,26 +3850,19 @@ def is_isomorphic(T1, T2):
     if T1.shape != T2.shape:
         raise ValueError('T1 and T2 must have the same number of elements.')
 
-    def py_is_isomorphic(T1, T2):
-        d1 = {}
-        d2 = {}
-        for t1, t2 in zip(T1, T2):
-            if t1 in d1:
-                if t2 not in d2:
-                    return False
-                if d1[t1] != t2 or d2[t2] != t1:
-                    return False
-            elif t2 in d2:
-                return False
-            else:
-                d1[t1] = t2
-                d2[t2] = t1
-        return True
+    # For each pair of (i, j) indices, test that
+    # T1[i] == T1[j] <--> T2[i] == T2[j]
+
+    # O(n*log(n)) algorithm.
+    # It is also possible to write a O(n) algorithm on top of unique_all(),
+    # but in practice it's much slower even for large clusters.
+    idx = xp.argsort(T1)
+    T1 = xp.take(T1, idx)
+    T2 = xp.take(T2, idx)
+    changes1 = T1[:-1] == T1[1:]
+    changes2 = T2[:-1] == T2[1:]
+    return xp.all(changes1 == changes2)
```

Comment:
Changes from aligning to `is_valid_linkage` and `is_valid_im` to aligning to `is_monotonic`, which feels much more natural.
Note that this changes the return type on numpy from `bool` to `np.bool_`. It shouldn't be a problem worth worrying about IMHO?

Reply:
> This PR changes the return value

Can you clarify--this is a change in the public API return value, even when using NumPy? It is getting a bit hard for me to track/review all this is going on, but that kind of change without i.e., deprecation/some kind of transition would give me pause.

Perhaps a concrete example of the change when using NumPy might suit for reviewers with limited time.

Reply:
Yes, this changes the return type on numpy from `bool` to `np.bool_`.
I could add a switch that reverts _just numpy_ to `bool`, but it feels like an antipattern?

Reply:
I'm just saying we don't normally make backwards-incompatible changes because of intuition (you probably know this)--normally they need to be justified as bug fixes or go through some kind of process. That specific example sounds fairly minor, but hard to judge as a non-consumer of the functionality.

Anyway, just making sure this gets discussed before merge.

Reply:
I've changed it to return `bool` when SCIPY_ARRAY_API is not defined.

You still have a breaking change on `SCIPY_ARRAY_API=1`; most notably `is_isomorphic(a, b) is True` and `is False` (which are commonly regarded as antipatterns) will always return False.

Would this be sufficient? If not, how would you suggest implementing a deprecation cycle for a return type?